File: src/skillguard/sdk/manifest.py
"""Skill manifest schema - explicit permission declarations."""

from __future__ import annotations

from enum import StrEnum
from pathlib import Path
from typing import Literal

import yaml
from pydantic import BaseModel, Field, field_validator


class HttpMethod(StrEnum):
    """Allowed HTTP methods."""
    GET = "GET"
    POST = "POST"
    PUT = "PUT"
    PATCH = "PATCH"
    DELETE = "DELETE"
    HEAD = "HEAD"
    OPTIONS = "OPTIONS"


class FilesystemAccess(StrEnum):
    """Filesystem access levels."""
    READ = "read"
    WRITE = "write"
    EXECUTE = "execute"


class PermissionLevel(StrEnum):
    """Overall permission levels for skills."""
    MINIMAL = "minimal"
    RESTRICTED = "restricted"
    STANDARD = "standard"
    PRIVILEGED = "privileged"


class NetworkPermission(BaseModel):
    """Network access permission for a specific domain."""

    domain: str = Field(..., description="Domain or IP to allow access to")
    methods: list[HttpMethod] = Field(
        default=[HttpMethod.GET],
        description="Allowed HTTP methods"
    )
    ports: list[int] = Field(
        default=[443, 80],
        description="Allowed ports"
    )

    @field_validator("domain")
    @classmethod
    def validate_domain(cls, v: str) -> str:
        if not v or v.isspace():
            raise ValueError("Domain cannot be empty")
        return v.lower().strip()


class FilesystemPermission(BaseModel):
    """Filesystem access permission for a specific path pattern."""

    path: str = Field(..., description="Path pattern (supports ${WORKSPACE} variable)")
    access: list[FilesystemAccess] = Field(
        default=[FilesystemAccess.READ],
        description="Allowed access types"
    )

    @field_validator("path")
    @classmethod
    def validate_path(cls, v: str) -> str:
        if not v or v.isspace():
            raise ValueError("Path cannot be empty")
        return v.strip()


class EnvironmentPermission(BaseModel):
    """Environment variable access permission."""

    name: str = Field(..., description="Environment variable name")
    required: bool = Field(default=False, description="Whether the variable is required")
    sensitive: bool = Field(default=False, description="Whether the variable contains sensitive data")


class Permission(BaseModel):
    """Complete permission manifest for a skill."""

    network: list[NetworkPermission] = Field(
        default_factory=list,
        description="Network access permissions"
    )
    filesystem: list[FilesystemPermission] = Field(
        default_factory=list,
        description="Filesystem access permissions"
    )
    environment: list[EnvironmentPermission] = Field(
        default_factory=list,
        description="Environment variable permissions"
    )
    subprocess: bool = Field(
        default=False,
        description="Whether subprocess execution is allowed"
    )
    subprocess_allowlist: list[str] = Field(
        default_factory=list,
        description="Allowed subprocess commands (if subprocess is True)"
    )

    @property
    def level(self) -> PermissionLevel:
        """Compute the effective permission level."""
        if not self.network and not self.filesystem and not self.subprocess:
            return PermissionLevel.MINIMAL
        if self.subprocess and not self.subprocess_allowlist:
            return PermissionLevel.PRIVILEGED
        if self.subprocess:
            return PermissionLevel.STANDARD
        return PermissionLevel.RESTRICTED


class AdapterConfig(BaseModel):
    """Framework adapter compatibility configuration."""

    openclaw: str | None = Field(default=None, description="OpenClaw version constraint")
    langchain: str | None = Field(default=None, description="LangChain version constraint")
    crewai: str | None = Field(default=None, description="CrewAI version constraint")
    mcp: str | None = Field(default=None, description="MCP version constraint")
    autogpt: str | None = Field(default=None, description="AutoGPT version constraint")


class BuildConfig(BaseModel):
    """Build configuration for reproducible builds."""

    reproducible: bool = Field(default=True, description="Whether builds should be reproducible")
    base: str = Field(
        default="skillguard/python:3.11-minimal",
        description="Base image for building"
    )
    python_version: str = Field(default="3.11", description="Python version to use")
    dependencies: list[str] = Field(
        default_factory=list,
        description="Additional pip dependencies"
    )


class SecurityMetadata(BaseModel):
    """Security audit metadata."""

    audit_date: str | None = Field(default=None, description="Date of last security audit")
    auditor: str | None = Field(default=None, description="Entity that performed the audit")
    slsa_level: Literal[0, 1, 2, 3, 4] = Field(
        default=0,
        description="SLSA compliance level"
    )
    cve_scan_date: str | None = Field(default=None, description="Date of last CVE scan")
    known_vulnerabilities: list[str] = Field(
        default_factory=list,
        description="Known CVEs (should be empty for verified skills)"
    )


class SkillAction(BaseModel):
    """A single action that the skill can perform."""

    name: str = Field(..., description="Action name")
    description: str = Field(..., description="Human-readable description")
    parameters: dict[str, dict] = Field(
        default_factory=dict,
        description="JSON Schema for action parameters"
    )
    returns: dict = Field(
        default_factory=dict,
        description="JSON Schema for return value"
    )


class SkillManifest(BaseModel):
    """Complete skill manifest - the security contract for a skill."""

    name: str = Field(..., description="Skill name (lowercase, alphanumeric, hyphens)")
    version: str = Field(..., description="Semantic version")
    description: str = Field(..., description="Human-readable description")
    author: str = Field(..., description="Author or organization")
    license: str = Field(default="Apache-2.0", description="SPDX license identifier")
    homepage: str | None = Field(default=None, description="Project homepage URL")
    repository: str | None = Field(default=None, description="Source repository URL")

    permissions: Permission = Field(
        default_factory=Permission,
        description="Permission manifest"
    )
    adapters: AdapterConfig = Field(
        default_factory=AdapterConfig,
        description="Framework adapter compatibility"
    )
    build: BuildConfig = Field(
        default_factory=BuildConfig,
        description="Build configuration"
    )
    security: SecurityMetadata = Field(
        default_factory=SecurityMetadata,
        description="Security metadata"
    )

    actions: list[SkillAction] = Field(
        default_factory=list,
        description="Actions this skill can perform"
    )

    keywords: list[str] = Field(
        default_factory=list,
        description="Keywords for discovery"
    )

    @field_validator("name")
    @classmethod
    def validate_name(cls, v: str) -> str:
        import re
        if not re.match(r"^[a-z][a-z0-9-]*[a-z0-9]$", v) and len(v) > 1 and not re.match(r"^[a-z]$", v):
            raise ValueError(
                "Name must be lowercase, start with a letter, "
                "and contain only letters, numbers, and hyphens"
            )
        return v

    @field_validator("version")
    @classmethod
    def validate_version(cls, v: str) -> str:
        from packaging.version import Version
        try:
            Version(v)
        except Exception as e:
            raise ValueError(f"Invalid semantic version: {v}") from e
        return v

    @classmethod
    def from_yaml(cls, path: Path | str) -> SkillManifest:
        """Load a manifest from a YAML file."""
        path = Path(path)
        if not path.exists():
            raise FileNotFoundError(f"Manifest not found: {path}")

        with open(path) as f:
            data = yaml.safe_load(f)

        return cls.model_validate(data)

    def to_yaml(self, path: Path | str) -> None:
        """Write the manifest to a YAML file."""
        path = Path(path)
        path.parent.mkdir(parents=True, exist_ok=True)

        with open(path, "w") as f:
            yaml.dump(
                self.model_dump(exclude_none=True, exclude_defaults=True),
                f,
                default_flow_style=False,
                sort_keys=False,
            )

    def to_yaml_string(self) -> str:
        """Serialize the manifest to a YAML string."""
        return yaml.dump(
            self.model_dump(exclude_none=True),
            default_flow_style=False,
            sort_keys=False,
        )

File: rust/crates/skillguard-core/src/manifest.rs
use crate::action::SkillAction;
use crate::error::{Result, SkillGuardError};
use crate::permission::Permission;
use crate::validate::SkillName;
use serde::{Deserialize, Serialize};
use std::path::Path;

/// Adapter framework version constraints.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct AdapterConfig {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub openclaw: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub langchain: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub crewai: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mcp: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub autogpt: Option<String>,
}

/// Build configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BuildConfig {
    #[serde(default = "default_true")]
    pub reproducible: bool,
    #[serde(default = "default_base_image")]
    pub base: String,
    #[serde(default = "default_python_version")]
    pub python_version: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub dependencies: Vec<String>,
}

fn default_true() -> bool {
    true
}

fn default_base_image() -> String {
    "skillguard/python:3.11-minimal".into()
}

fn default_python_version() -> String {
    "3.11".into()
}

impl Default for BuildConfig {
    fn default() -> Self {
        Self {
            reproducible: true,
            base: default_base_image(),
            python_version: default_python_version(),
            dependencies: Vec::new(),
        }
    }
}

/// Security metadata.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct SecurityMetadata {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub audit_date: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auditor: Option<String>,
    #[serde(default)]
    pub slsa_level: u8,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cve_scan_date: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub known_vulnerabilities: Vec<String>,
}

/// The complete skill manifest — the security contract.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkillManifest {
    pub name: String,
    pub version: String,
    pub description: String,
    pub author: String,
    #[serde(default = "default_license")]
    pub license: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub homepage: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub repository: Option<String>,
    #[serde(default)]
    pub permissions: Permission,
    #[serde(default, skip_serializing_if = "is_default_adapter")]
    pub adapters: AdapterConfig,
    #[serde(default)]
    pub build: BuildConfig,
    #[serde(default)]
    pub security: SecurityMetadata,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub actions: Vec<SkillAction>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub keywords: Vec<String>,
}

fn default_license() -> String {
    "Apache-2.0".into()
}

fn is_default_adapter(config: &AdapterConfig) -> bool {
    config.openclaw.is_none()
        && config.langchain.is_none()
        && config.crewai.is_none()
        && config.mcp.is_none()
        && config.autogpt.is_none()
}

impl SkillManifest {
    /// Load and validate a manifest from a YAML file.
    pub fn from_yaml_file(path: impl AsRef<Path>) -> Result<Self> {
        let content = std::fs::read_to_string(path.as_ref())?;
        Self::from_yaml_str(&content)
    }

    /// Parse and validate a manifest from a YAML string.
    pub fn from_yaml_str(yaml: &str) -> Result<Self> {
        let manifest: Self = serde_yaml::from_str(yaml)?;
        manifest.validate()?;
        Ok(manifest)
    }

    /// Write the manifest to a YAML file.
    pub fn to_yaml_file(&self, path: impl AsRef<Path>) -> Result<()> {
        let yaml = serde_yaml::to_string(self)?;
        std::fs::write(path.as_ref(), yaml)?;
        Ok(())
    }

    /// Serialize to a YAML string.
    pub fn to_yaml_string(&self) -> Result<String> {
        Ok(serde_yaml::to_string(self)?)
    }

    /// Validate the manifest fields.
    pub fn validate(&self) -> Result<()> {
        // Validate skill name
        SkillName::new(&self.name)?;

        // Validate version
        semver::Version::parse(&self.version).map_err(|e| {
            SkillGuardError::InvalidManifest(format!("Invalid version '{}': {}", self.version, e))
        })?;

        // Validate SLSA level
        if self.security.slsa_level > 4 {
            return Err(SkillGuardError::InvalidManifest(format!(
                "SLSA level must be 0-4, got {}",
                self.security.slsa_level
            )));
        }

        // Validate actions have names
        for action in &self.actions {
            if action.name.is_empty() {
                return Err(SkillGuardError::InvalidManifest(
                    "Action name cannot be empty".into(),
                ));
            }
        }

        // Validate network permissions
        for net in &self.permissions.network {
            if net.domain.is_empty() {
                return Err(SkillGuardError::InvalidManifest(
                    "Network permission domain cannot be empty".into(),
                ));
            }
        }

        Ok(())
    }

    /// Find an action by name.
    pub fn find_action(&self, name: &str) -> Option<&SkillAction> {
        self.actions.iter().find(|a| a.name == name)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_minimal_manifest() {
        let yaml = r#"
name: test-skill
version: 1.0.0
description: A test skill
author: test-author
"#;
        let manifest = SkillManifest::from_yaml_str(yaml).unwrap();
        assert_eq!(manifest.name, "test-skill");
        assert_eq!(manifest.version, "1.0.0");
        assert_eq!(manifest.license, "Apache-2.0");
        assert!(manifest.actions.is_empty());
    }

    #[test]
    fn test_invalid_name() {
        let yaml = r#"
name: Invalid_Name
version: 1.0.0
description: Bad
author: test
"#;
        assert!(SkillManifest::from_yaml_str(yaml).is_err());
    }

    #[test]
    fn test_invalid_version() {
        let yaml = r#"
name: test-skill
version: not-semver
description: Bad
author: test
"#;
        assert!(SkillManifest::from_yaml_str(yaml).is_err());
    }

    #[test]
    fn test_roundtrip_yaml() {
        let yaml = r#"
name: test-skill
version: 1.0.0
description: A test skill
author: test-author
permissions:
  network:
    - domain: example.com
      methods: [GET, POST]
      ports: [443]
  filesystem:
    - path: "${WORKSPACE}/**"
      access: [read, write]
actions:
  - name: do-thing
    description: Does a thing
keywords:
  - test
"#;
        let manifest = SkillManifest::from_yaml_str(yaml).unwrap();
        let output = manifest.to_yaml_string().unwrap();
        let reparsed = SkillManifest::from_yaml_str(&output).unwrap();
        assert_eq!(manifest.name, reparsed.name);
        assert_eq!(manifest.actions.len(), reparsed.actions.len());
    }
}
